src/shallow.ts

```diff
@@ -59,3 +61,17 @@ export default ((objA, objB) => {
   }
   return shallow(objA, objB)
 }) as typeof shallow
+
+export function useShallow<S, U>(selector: (state: S) => U): (state: S) => U {
+  const prev = useRef<U>()
+
+  return useMemo(
+    () => (state) => {
+      const next = selector(state)
+      return shallow(prev.current, next)
+        ? (prev.current as U)
+        : (prev.current = next)
+    },
+    [selector]
```

Comment:
I don't think we need useMemo nor useCallback for this, because Zustand doesn't require stable selector.

Do you prefer `useRef`? I think we should use useMemo like https://github.com/pmndrs/zustand/discussions/1937#discussioncomment-6974554 , because React may prohibit updating ref during render in the future (L73 is usually not allowed).

Reply:
Version 1 of https://github.com/pmndrs/zustand/discussions/1937#discussioncomment-6974554 requires useCallback to use inline selectors.

Version 2 of https://github.com/pmndrs/zustand/discussions/1937#discussioncomment-6974554 works but we open the door to potential stale closure issues.


If there are no better alternatives version 2 is better.

Are there less cumbersome alternatives.

---

I wish we could use [useEvent or something similar](https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md) during the render phase :/

Reply:
Oops, that's my bad. I forgot about it. Let's use `useRef` then.

I think you can remove useMemo from your implementation.

Reply:
> It's not necessary because the caller of useShallow can always wrap the selector with useCallabck.

We never need useCallback, it doesn't improve anything.